src/cudadecoder/cuda-decoder-kernels-utils.h

```diff
@@ -0,0 +1,215 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_CUDA_DECODER_CUDA_DECODER_KERNELS_UTILS_H_
+#define KALDI_CUDA_DECODER_CUDA_DECODER_KERNELS_UTILS_H_
+
+// NO_KEY == -1 is ok, because all keys will be >= 0 (FST states)
+#define KALDI_CUDA_DECODER_HASHMAP_NO_KEY -1
+#define KALDI_CUDA_DECODER_HASHMAP_NO_VAL                 \
+  {                                                       \
+    KALDI_CUDA_DECODER_HASHMAP_NO_KEY, 0, { INT_MAX, -1 } \
+  }
+
+#include "util/stl-utils.h"
+
+namespace kaldi {
+namespace cuda_decoder {
+
+// MinPlus and PlusPlus
+// int2 operators used in Scan or Reduce operations
+struct MinPlus {
+  __device__ int2 operator()(const int2 &a, const int2 &b) const {
+    int2 c;
+    c.x = min(a.x, b.x);
+    c.y = a.y + b.y;
+    return c;
+  }
+};
+struct PlusPlus {
+  __device__ int2 operator()(const int2 &a, const int2 &b) const {
+    int2 c;
+    c.x = a.x + b.x;
+    c.y = a.y + b.y;
+    return c;
+  }
+};
+
+// 1:1 Conversion float <---> sortable int
+// We convert floats to sortable ints in order
+// to use native atomics operation, which are
+// way faster than looping over atomicCAS
+__device__ __forceinline__ int32 floatToOrderedInt(float floatVal) {
+  int32 intVal = __float_as_int(floatVal);
+  return (intVal >= 0) ? intVal : intVal ^ 0x7FFFFFFF;
+}
+
+__device__ __forceinline__ float orderedIntToFloat(int32 intVal) {
+  return __int_as_float((intVal >= 0) ? intVal : intVal ^ 0x7FFFFFFF);
+}
+
+// binsearch_maxle (device)
+// With L=[all indexes low<=i<=high such as vec[i]<= val]
+// binsearch_maxle returns max(L)
+// the array vec must be sorted
+// Finds that value using a binary search
+__device__ __forceinline__ int32 binsearch_maxle(const int32 *vec,
+                                                 const int32 val, int32 low,
+                                                 int32 high) {
+  while (true) {
+    if (low == high) return low;  // we know it exists
+    if ((low + 1) == high) return (vec[high] <= val) ? high : low;
+
+    int32 mid = low + (high - low) / 2;
+
+    if (vec[mid] > val)
+      high = mid - 1;
+    else
+      low = mid;
+  }
+}
+
+// Atomic operations on int2 (device)
+// atomicAddI2, atomicMinI2, atomicSubI2
+//
+// union used
+union UInt64UnionInt2 {
+  int2 i2;
+  unsigned long long int ull;
+};
+
+__device__ __inline__ int2 atomicAddI2(int2 *ptr, int2 val) {
+  unsigned long long int *ptr64 =
+      reinterpret_cast<unsigned long long int *>(ptr);
+  UInt64UnionInt2 uval, uold;
+  uval.i2 = val;
+  uold.ull = atomicAdd(ptr64, uval.ull);
+  return uold.i2;
+}
+
+// We should switch to native atom64 on atomicMinI2 and atomicSubI2
```

Comment:
Yes, why isn't this using [atomicMin](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicmin) overloaded for `unsigned long long int`? It seems like it would be way faster that way. Do you really need to support compute capability 3.0?

Reply:
Yes this is in the list. For 3.0 I guess it is still supported in the Kaldi makefiles (we can keep the duplicate and add a #if __CUDA_ARCH__)